src/details/second_order_lowpass.cpp

```diff
@@ -3,34 +3,41 @@
 
 namespace path_tracking_pid::details
 {
-namespace
+SecondOrderLowpass::SecondOrderLowpass(double cutoff, double damping)
+: cutoff_(cutoff), damping_(damping)
 {
-// Used in filter calculations. Default 1.0 corresponds to a cutoff frequency at 1/4 of the sample rate.
-constexpr double cutoff = 1.;
-
-}  // namespace
+}
 
-void SecondOrderLowpass::push(double value)
+void SecondOrderLowpass::configure(double cutoff, double damping)
 {
-  errors_.push(value);
-
-  filtered_errors_.push(
-    (1 / (1 + cutoff * cutoff + M_SQRT2 * cutoff)) *
-    (errors_.at<2>() + 2 * errors_.at<1>() + errors_.at<0>() -
-     (cutoff * cutoff - M_SQRT2 * cutoff + 1) * filtered_errors_.at<1>() -
-     (-2 * cutoff * cutoff + 2) * filtered_errors_.at<0>()));
+  cutoff_ = cutoff;
+  damping_ = damping;
 }
 
-void SecondOrderLowpass::reset()
+double SecondOrderLowpass::filter(double u, double step_size)
 {
-  errors_.reset();
-  filtered_errors_.reset();
+  // save history
+  u_.push(u);
+  y_.push(u);  // increase index so the math below looks correct
+
+  if (cutoff_ == 0) {
+    return u;
+  }
+
+  auto c = cutoff_;
+  auto d = damping_;
+  auto T = step_size;
+  auto a = 2 * M_PI * c;
+  auto b = T * a;
+  y_[0] = ((pow(b, 2)) * u_[0] + (2 * pow(b, 2)) * u_[1] + (pow(b, 2)) * u_[2] -
+           (2 * pow(b, 2) - 8) * y_[1] - (pow(b, 2) - 4 * T * a * d + 4) * y_[2]) /
+          (pow(b, 2) + 4 * T * a * d + 4);
```

Comment:
@MCFurry can you verify that I've made no mistakes in the math for the filter?

This is how I've come to these values:
https://github.com/nobleo/path_tracking_pid/blob/fix/lowpass/doc/second_order_lowpass_tustin.ipynb